File: flutter_zoom_videosdk-2.4.0/ios/Classes/FlutterZoomVideoSdkVideoHelper.m
#import "FlutterZoomVideoSdkVideoHelper.h"
#import "FlutterZoomVideoSdkUser.h"
#import "FlutterZoomVideoSdkCameraDevice.h"
#import "JSONConvert.h"

@implementation FlutterZoomVideoSdkVideoHelper

- (ZoomVideoSDKVideoHelper *)getVideoHelper {
 ZoomVideoSDKVideoHelper* videoHelper = nil;
 @try {
     videoHelper = [[ZoomVideoSDK shareInstance] getVideoHelper];
     if (videoHelper == nil) {
         NSException *e = [NSException exceptionWithName:@"NoVideoHelperFound" reason:@"No Video Helper Found" userInfo:nil];
         @throw e;
     }
 } @catch (NSException *e) {
     NSLog(@"%@ - %@", e.name, e.reason);
 }
 return videoHelper;
}

-(void) startVideo: (FlutterResult) result {
 dispatch_async(dispatch_get_main_queue(), ^{
     result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getVideoHelper] startVideo])]);
 });
}

-(void) stopVideo: (FlutterResult) result {
 dispatch_async(dispatch_get_main_queue(), ^{
     result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getVideoHelper] stopVideo])]);
 });
}

-(void) rotateMyVideo: (FlutterMethodCall *)call withResult:(FlutterResult) result {
 dispatch_async(dispatch_get_main_queue(), ^{
    if ([[self getVideoHelper] rotateMyVideo:(UIDeviceOrientation) call.arguments[@"rotation"]]) {
        result(@YES);
    } else {
        result(@NO);
    }
 });
}

-(void) switchCamera: (FlutterMethodCall *)call withResult:(FlutterResult) result {
 dispatch_async(dispatch_get_main_queue(), ^{
     id deviceId = call.arguments[@"deviceId"];
     if (deviceId == nil || [deviceId isKindOfClass:[NSNull class]] || [deviceId length] == 0) {
         [[self getVideoHelper] switchCamera];
         result(@YES);
     } else {
         NSArray *cameraList = [[[ZoomVideoSDK shareInstance] getVideoHelper] getCameraDeviceList];
         for (ZoomVideoSDKCameraDevice *device in cameraList) {
             if ([device.deviceId isEqualToString: call.arguments[@"deviceId"]]) {
                 if ([[[ZoomVideoSDK shareInstance] getVideoHelper] switchCamera:call.arguments[@"deviceId"]]) {
                     result(@YES);
                 } else {
                     result(@NO);
                 }
                 return;
             }
         }
         result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @(Errors_Invalid_Parameter)]
                                    message:@"Camera not found"
                                    details:nil]);
     }
 });
}

-(void) getCameraDeviceList: (FlutterResult) result {
 dispatch_async(dispatch_get_main_queue(), ^{
     NSArray <ZoomVideoSDKCameraDevice*>* cameraDeviceList = [[self getVideoHelper] getCameraDeviceList];
     result([FlutterZoomVideoSdkCameraDevice mapCameraDeviceArray: cameraDeviceList]);
 });
}

-(void) mirrorMyVideo: (FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getVideoHelper] mirrorMyVideo: [call.arguments[@"enable"] boolValue]])]);
    });
}

-(void) isMyVideoMirrored: (FlutterResult) result {
    if ([[self getVideoHelper] isMyVideoMirrored]) {
        result(@YES);
    } else {
        result(@NO);
    }
}

-(void) enableOriginalAspectRatio: (FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        result(@([[self getVideoHelper] enableOriginalAspectRatio: [call.arguments[@"enabled"] boolValue]]));
    });
}

-(void) isOriginalAspectRatioEnabled: (FlutterResult) result {
    if ([[self getVideoHelper] isOriginalAspectRatioEnabled]) {
        result(@YES);
    } else {
        result(@NO);
    }
}

-(void) spotLightVideo: (FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString* userId = call.arguments[@"userId"];
        if ([userId length] == 0) {
            result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @(Errors_Invalid_Parameter)]
                                       message:@"UserId is empty"
                                       details:nil]);
        }
        ZoomVideoSDKUser *user = [FlutterZoomVideoSdkUser getUser:userId];
        if (user != nil) {
            result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getVideoHelper] spotLightVideo:user])]);
        } else {
            result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @(Errors_Invalid_Parameter)]
                                       message:@"User not found"
                                       details:nil]);
        }
    });
}

-(void) unSpotLightVideo: (FlutterMethodCall *)call withResult:(FlutterResult) result {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString* userId = call.arguments[@"userId"];
        if ([userId length] == 0) {
            result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @(Errors_Invalid_Parameter)]
                                       message:@"UserId is empty"
                                       details:nil]);
        }
        ZoomVideoSDKUser *user = [FlutterZoomVideoSdkUser getUser:userId];
        if (user != nil) {
            result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getVideoHelper] unSpotLightVideo:user])]);
        } else {
            result([FlutterError errorWithCode:[[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @(Errors_Invalid_Parameter)]
                                       message:@"User not found"
                                       details:nil]);
        }
        
    });
}

-(void) unSpotlightAllVideos: (FlutterResult) result {
 dispatch_async(dispatch_get_main_queue(), ^{
     result([[JSONConvert ZoomVideoSDKErrorValuesReversed] objectForKey: @([[self getVideoHelper] unSpotlightAllVideos])]);
 });
}

-(void) getSpotlightedVideoUserList: (FlutterResult) result {
 dispatch_async(dispatch_get_main_queue(), ^{
     result([FlutterZoomVideoSdkUser mapUserArray:[[self getVideoHelper] getSpotlightedVideoUserList]]);
 });
}

@end

File: flutter_zoom_videosdk-2.4.0/android/src/main/java/com/flutterzoom/videosdk/FlutterZoomVideoSdkVideoHelper.java
package com.flutterzoom.videosdk;

import android.app.Activity;

import androidx.annotation.NonNull;

import com.flutterzoom.videosdk.convert.FlutterZoomVideoSdkErrors;

import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import us.zoom.sdk.ZoomVideoSDK;
import us.zoom.sdk.ZoomVideoSDKCameraDevice;
import us.zoom.sdk.ZoomVideoSDKErrors;
import us.zoom.sdk.ZoomVideoSDKUser;
import us.zoom.sdk.ZoomVideoSDKVideoHelper;

public class FlutterZoomVideoSdkVideoHelper {

    private Activity activity;

    FlutterZoomVideoSdkVideoHelper(Activity activity) {
        this.activity = activity;
    }

    private ZoomVideoSDKVideoHelper getVideoHelper() {
        ZoomVideoSDKVideoHelper videoHelper = null;
        try {
            videoHelper = ZoomVideoSDK.getInstance().getVideoHelper();
            if (videoHelper == null) {
                throw new Exception("No Video Helper Found");
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return videoHelper;
    }

    public void getCameraList(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkCameraDevice.jsonCameraArray(getVideoHelper().getCameraList()));
            }
        });
    }

    public void getNumberOfCameras(@NonNull MethodChannel.Result result) {
        result.success(getVideoHelper().getNumberOfCameras());
    }

    public void rotateMyVideo(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> params = call.arguments();
        double rotation = (Double) params.get("rotation");

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(getVideoHelper().rotateMyVideo((int) rotation));
            }
        });
    }

    public void startVideo(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getVideoHelper().startVideo()));
            }
        });
    }

    public void stopVideo(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getVideoHelper().stopVideo()));
            }
        });
    }

    public void switchCamera(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> params = call.arguments();
        String deviceId = (String) params.get("deviceId");

        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                if (deviceId == null || deviceId.isEmpty()) {
                    result.success(getVideoHelper().switchCamera());
                } else {
                    ZoomVideoSDKCameraDevice device = getVideoHelper().getCameraList()
                            .stream()
                            .filter(c -> c.getDeviceId().equals(deviceId))
                            .findAny()
                            .orElse(null);
                    if (device != null) {
                        result.success(getVideoHelper().switchCamera(device));
                    } else {
                        result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter),
                                "Camera not found",
                                null);
                    }
                }
            }
        });
    }

    public void mirrorMyVideo(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> params = call.arguments();
        boolean enable = (boolean) params.get("enable");
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getVideoHelper().mirrorMyVideo(enable)));
            }
        });
    }

    public void isMyVideoMirrored(@NonNull MethodChannel.Result result) {
        result.success(getVideoHelper().isMyVideoMirrored());
    }

    public void isOriginalAspectRatioEnabled(@NonNull MethodChannel.Result result) {
        result.success(getVideoHelper().isOriginalAspectRatioEnabled());
    }

    public void enableOriginalAspectRatio(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> params = call.arguments();
        boolean enable = (boolean) params.get("enable");
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(getVideoHelper().enableOriginalAspectRatio(enable));
            }
        });
    }

    public void turnOnOrOffFlashlight(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> params = call.arguments();
        boolean isOn = (boolean) params.get("isOn");
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(getVideoHelper().turnOnOrOffFlashlight(isOn));
            }
        });
    }


    public void isSupportFlashlight(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(getVideoHelper().isSupportFlashlight());
            }
        });
    }

    public void isFlashlightOn(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(getVideoHelper().isFlashlightOn());
            }
        });
    }

    public void spotLightVideo(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> params = call.arguments();
        String userId = (String) params.get("userId");
        if (userId == null || userId.isEmpty()) {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "userId is null or empty", null);
        }
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                ZoomVideoSDKUser user = FlutterZoomVideoSdkUser.getUser(userId);
                if (user != null) {
                    result.success(FlutterZoomVideoSdkErrors.valueOf(getVideoHelper().spotLightVideo(user)));
                } else {
                    result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "User not found", null);
                }
            }
        });
    }

    public void unSpotLightVideo(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        Map<String, Object> params = call.arguments();
        String userId = (String) params.get("userId");
        if (userId == null || userId.isEmpty()) {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "userId is null or empty", null);
        }
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                ZoomVideoSDKUser user = FlutterZoomVideoSdkUser.getUser(userId);
                if (user != null) {
                    result.success(FlutterZoomVideoSdkErrors.valueOf(getVideoHelper().unSpotLightVideo(user)));
                } else {
                    result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Invalid_Parameter), "User not found", null);
                }
            }
        });
    }

    public void unSpotlightAllVideos(@NonNull MethodChannel.Result result) {
        activity.runOnUiThread(new Runnable() {
            @Override
            public void run() {
                result.success(FlutterZoomVideoSdkErrors.valueOf(getVideoHelper().unSpotlightAllVideos()));
            }
        });
    }

    public void getSpotlightedVideoUserList(@NonNull MethodChannel.Result result) {
        List<ZoomVideoSDKUser> spotlightedVideoUsers = getVideoHelper().getSpotlightedVideoUserList();

        if (spotlightedVideoUsers == null) {
            result.error(FlutterZoomVideoSdkErrors.valueOf(ZoomVideoSDKErrors.Errors_Spotlight_NotEnoughUsers), "spotlighted user list is null", null);
            return;
        }
        result.success(FlutterZoomVideoSdkUser.jsonUserArray(spotlightedVideoUsers));
    }
}